styx-common/src/main/java/com/spotify/styx/util/ShardedCounter.java

```diff
@@ -0,0 +1,364 @@
+   * that strongly consistently increment resp. decrement the counter referred to by counterId, and
+   * cause the transaction to fail to commit if the counter's associated limit is exceeded. Also
+   * spurious failures are possible.
+   *
+   * <p>Delta should be +/-1 for graceful behavior, due to how sharding is currently implemented.
+   * Updates with a larger delta are prone to spuriously fail even when the counter is not near to
+   * exceeding its limit. Failures are certain when delta >= limit / NUM_SHARDS + 1.
+   */
+  public void updateCounter(DatastoreReaderWriter transaction, String counterId, long delta) {
+    CounterSnapshot snapshot = getCounterSnapshot(counterId);
+    int shardIndex;
+    try {
+      shardIndex = snapshot.pickShardWithSpareCapacity(delta);
+    } catch (InvalidShardStateException e) {
+      try {
+        snapshot = refreshCounterSnapshot(counterId);
+        shardIndex = snapshot.pickShardWithSpareCapacity(delta);
+      } catch (InvalidShardStateException e1) {
+        LOG.warn(
+            "Invalid shard state for counter {} but cache has already been reloaded for counter. {}",
+            counterId, e1);
+        return;
+      }
+    }
+    final Key shardKey = datastore.newKeyFactory().setKind(KIND_COUNTER_SHARD)
+        .newKey(counterId + "-" + shardIndex);
+    final Entity shard = transaction.get(shardKey);
+
+    final long newCounterValue = shard.getLong(PROPERTY_SHARD_VALUE) + delta;
+    if (shard != null && Range.closed(0L, snapshot.shardCapacity(shardIndex))
+        .contains(newCounterValue) && newCounterValue <= getLimit(datastore, counterId)) {
+      transaction.put(Entity.newBuilder(shard)
+                          .set(PROPERTY_SHARD_VALUE,
+                               shard.getLong(PROPERTY_SHARD_VALUE) + delta)
+                          .build());
+    } else {
+      // TODO fail (rollback) the transaction
+      throw new ConcurrentModificationException(
+          "Chosen shard has no spare capacity anymoar, pls rollback kths");
+    }
+  }
+
+  /**
+   * Returns the current value of the counter referred to by counterId, a weakly consistent
+   * estimate. (May have not truly been the counter value at any point in time. Even a return value
+   * larger than the corresponding limit might be possible without error.)
+   */
+  public long getCounter(String counterId) {
+    final CounterSnapshot snapshot = getCounterSnapshot(counterId);
+    long result = 0;
+
+    for (long shard : snapshot.shards.values()) {
+      result += shard;
+    }
+    return result;
+  }
+
+  /**
+   * Delete counter by counterId. Deletes both counter shards and counter limit if it exists.
+   *
+   * <p>There may be inconsistencies between instances, as each stores an in-memory cache, whose
+   * state may differ from what's stored in Datastore. Only after a cache expiry will other instances
+   * notice that a counter has been deleted.
```

Comment:
The other methods aren't in fact designed to notice such condition. They will just go and initialize the counter as if it was just coming into existence. Maybe be can say just that is that it's up to the user to ensure there's no instances expecting the counter to exist while deleteCounter is ongoing...

Reply:
True true